### DISCUSSION TITLE
Build database and recursively include dependencies

### DISCUSSION THREAD

**Discussion by mtardy** [github/securitylab#698] [Q&A] [answered]
I would like to build a database for kubelet, in the Kubernetes monorepo: https://github.com/kubernetes/kubernetes.
It partially worked with: `codeql database create kubelet-db --language=go --command='make kubelet'` but my issue is that it's missing the code of its local dependencies. For example, [this package](https://github.com/kubernetes/kubernetes/tree/master/pkg/volume/util/subpath) is correctly imported because it's built into the `kubelet` binary but its local dependency [k8s.io/mount-utils](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/subpath/subpath_linux.go#L33) (which is in a different go module) is not included (available [here](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/mount-utils)).

It's unfortunate because a non-negligible portion of the code actually happens in these modules.

Is there a way to include these dependencies into the build?

EDIT: I guess I should have asked on [github/codeql](https://github.com/github/codeql/discussions) but I cannot remove nor transfer this discussion...

**Comment by smowton** [comment, marked as answer]
you can give `--command` more than once -- does it help to add extra build commands that build the dependency you'd like to extract?

**Reply by mtardy** [reply]
I will try but it's quite problematic because for example if I'm searching for datapath between source and `exec.Command` sinks I cannot be sure that I cover all of `exec.Command` without trying to build specific dependencies by hand. There is no way to recursively build the modules that are called in the code? It would be too big?

**Reply by mtardy** [reply]
It works 😄 ! Which is super nice but is there a systematic way of building all the modules recursively automatically?

**Reply by smowton** [reply]
I imagine a plain old `make` builds everything? But if you want to figure out what restricted build will build less than everything but at least what's reachable from a given target, then no that isn't automated; you have to cook up a build command that touches the needed code.
